### DISCUSSION TITLE
[API Proposal]: Record Default Constructor Accessibility

### DISCUSSION THREAD

**Discussion by TheBrambleShark** [dotnet/csharplang#6310] [Language Ideas]
### Background and motivation

Currently with record types, whether record classes or record structs, the default constructor is always public. There are times where non-public constructors are wanted, however to do so, users must implement the properties and constructor manually.

There are certainly cases where this is better, but it removes the utility of the default constructor for record types.

```cs
// old
public sealed record class Person(string Name, int Age);

// new
public sealed record class Person
{
    public string Name { get; }
    public int Age { get; }
    
    private Person(string name, int age)
    {
        Name = name;
        Age = age;
     }
 }
 ```

### API Proposal

I would like to propose an additional syntax for record types, as follows:

```csharp
public sealed record class Person private(string Name, int Age);
public readonly record struct Person protected internal(string Name, int Age);
```
This would compile into a class or struct as normal, but the specified accessibility modifier would be applied to the generated constructor.

Necessarily, the compiler will need instructions that if an accessibility modifier is not applied, then it should be assumed to be `public` in order to maintain backwards compatibility with existing record types.

Optionally, an analyzer could also be created to recommend removing the `public` modifier for simplification reasons, 

### API Usage

The proposal and the usage are pretty much the same thing.

### Alternative Designs

The placement of the private modifier is based off of Kotlin's class specification. Here is an example of theirs:
```kt
class Foo private constructor(Bar bar) { }
```
That said, we are not beholden to any design because of another language's decisions, so if a different placement is desired for different reasons, I take no issue with this.

### Risks

So long as we make the modifier optional and specify that no modifier is the same as public, there are no risks to this change.

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

**Comment by Xyncgas** [comment]
There is a problem because it's a record it needs a copy constructor that's why it's public so you can do
``` var MyRecordB = new Rcr (MyRecordA); ```
So people don't have to write copy constructors

Another thing is one constructors should always be public because that's the only way to create an object imo but I have no idea how kotlin can make their constructor private

**Reply by Xyncgas** [reply]
Maybe I am wrong

**Comment by huoyaoyuan** [comment]
What would you get for the record without public constructor? Only `ToString` and equality members?

**Reply by Xyncgas** [reply]
I supposed you can still make the with keyword work, that means there will always be a hidden default public constructor even you make it private every where else, or maybe the with keyword doesn't use any constructors

**Reply by TheBrambleShark** [reply]
I think allowing `with` to work is fine. We could probably even make the `Foo(Foo)` constructor public without issue as well.

One example case where this may be useful if you have an API that returns a type that shouldn't ever be created by the consumer. You could have an internal constructor so the nuget library which handles API interaction can create the type when deserializing the response, but there is no way for the consumer themselves to make one. Cloning an existing one shouldn't be an issue, nor would modifying the values (with the `with` keyword), but I'm not sure what that would afford them in that scenario.

**Reply by Xyncgas** [reply]
I do it lot in my codes too, sometimes I need to give you something that's black and shaped like a box and it can neither be created or destoryed, I created it in a factory, followed the promise through inherited interface or classes, that it's something you need, and sometimes if what I am giving to them doesn't need to be cast back letter so I can use it, I just go straight up construct a class, while you can ask me to do so by giving me another promise (e.g ``` public abstract static MyRequiredT (); ```) in a interface / abstract class that the factory will provide the implementations

**Comment by theunrepentantgeek** [comment]
> I supposed you can still make the with keyword work, that means there will always be a hidden default public constructor even you make it private every where else, or maybe the with keyword doesn't use any constructors

One common scenario where a private constructor is used is when a fixed set of instances is required but a regular enum isn't sufficient - say in the case of representing suits from a deck of cards. Such code typically expects there to be only the known instances (say, four). If the **with** operator allowed additional instances to be created, even though the constructor is private, would break such code.

Another scenario is where more control is desired than a public constructor affords, so a factory method is used instead. Such a factory method might do _pooling_, to reuse instances, or it might do _uniquing_, to reduce heap contention by reusing identical values. Again, a **with** operator that allowed additional instances to be created despite the constructor being private would be a breaking change.

Making the constructor private is expressing intent - it's saying _I want to tightly control which instances are created_. 

If records gain private constructors, the **with** keyword should _only_ work in contexts where the private constructor is already visible and accessible, nowhere else.

**Reply by TheBrambleShark** [reply]
You bring up a fair point. That said, the scenario you bring up is exactly how the `with` keyword behaves today. Consider the following example:
[SharpLab link to see it in action](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWACgsBmAAmQFMBjAe2QBNzdyAFK5AZzpmIG9i5YSwpY8ABnIA5CGCrl+5AOZUEAbnKwoG8gF8hIslpgJyAQVWKVazdt0GiI8oeHHxANkpUIjXgBsAT3ZOHhhyACk6BQBechgqAHcQ7l4ACgAiKKoMuHIcABYASnVXFycRAAdkKAA3CAQFDlSYNPEpGDkqPNgzCFUissEK52FZeXI4zvlSkdHLWPJ+qlnnR0didxwWPAB2ATLcHCGy53rkci5G2sXmsIA6bPJEnQALU5Fh0e+ZLsnyDIAZWuOQ+wj0qx+4gAnGkrlQbvdxlQSmV1kA)
```cs
using System;
public record class Person
{
    public string Name { get; init; }
    public int Age { get; init; }
    
    public static readonly Person Joe = new Person("Joe", 24);
    
    private Person(string name, int age)
    {
        Name = name;
        Age = age;
    }
}

public class Program
{
    public static void Main()
    {
        var steve = Person.Joe with
        {
            Name = "Steve"
        };
        Console.WriteLine(steve);
    }
}
```
In this case, the code compiles and "Steve" is written to the console, despite the fact that the constructor for `Person` is private.

As such, the question becomes whether this behavior should change when the visibility of the default constructor is changed. Of course, if we did so, this would be a breaking change (though I don't think people should be using `with` when the constructor is private anyways).

**Reply by Xyncgas** [reply]
Why is this happening ! Oh I think they made a hidden public constructor for you anyways since you opt-in to use record type

**Reply by TheBrambleShark** [reply]
> Why is this happening ! Oh I think they made a hidden public constructor for you anyways since you opt-in to use record type

Yup! It's called the "copy constructor." Here's the decompiled C# code:

```cs
public class Program
{
    public static void Main()
    {
        Person person = Person.Joe.<Clone>$();
        person.Name = "Steve";
        Console.WriteLine(person.Name);
    }
}
```
